src/Microsoft.AspNetCore.Authentication.Impl/AuthenticationHandler.cs

```diff
@@ -0,0 +1,229 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Text.Encodings.Web;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Internal;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Authentication{
+
+    public abstract class AuthenticationHandler<TOptions> : IAuthenticationHandler where TOptions : AuthenticationSchemeOptions, new()
+    {
+        private Task<AuthenticateResult> _authenticateTask;
+
+        public AuthenticationScheme Scheme { get; private set; }
+        public TOptions Options { get; private set; }
+        protected HttpContext Context { get; private set; }
+        protected PathString OriginalPathBase => Context.Features.Get<IAuthenticationFeature>()?.OriginalPathBase ?? Request.PathBase;
+
+        protected PathString OriginalPath => Context.Features.Get<IAuthenticationFeature>()?.OriginalPath ?? Request.Path;
+
+        protected ILogger Logger { get; }
+
+        protected UrlEncoder UrlEncoder { get; }
+
+        protected ISystemClock Clock { get; }
+
+        protected IOptionsFactory<TOptions> OptionsFactory { get; }
+
+        /// <summary>
+        /// The handler calls methods on the events which give the application control at certain points where processing is occurring. 
+        /// If it is not provided a default instance is supplied which does nothing when the methods are called.
+        /// </summary>
+        protected virtual object Events { get; set; }
```

Comment:
Are you sure every event isn't overridable?  Isn't that the only purpose for the IXyzEvents to exist?  basically its the list of virtual methods that would exist if you could plug in your own handler.